### ISSUE TITLE
Patch: Large file support on Unix systems

### ISSUE THREAD

**Issue by selmf** [zeniko/unarr#3]
The current file_seek implementation in common/stream.c supports large files (>2GB) on Windows/Visual Studio only and autofails on other systems. To implement large file support on Unix, the POSIX equivalents to fseeki64 and ftelli64 have to be used.

This patch adds large file support for Unix systems using the fseeko and ftello functions while keeping the old behaviour intact on other systems:

https://gist.github.com/selmf/d73617dfc2215d4bf01c

This should work fine on 64 bit systems, but 32 bit systems will need one more modification. Quoting the fseeko manpage:

 "On  some  architectures,  both off_t and long are 32-bit types, but defining _FILE_OFFSET_BITS with
  the value 64 (before including any header files) will turn off_t into a 64-bit type"

A good way to achieve this without header troubles is adding "-D_FILE_OFFSET_BITS=64" to the compiler flags.

I hope this provides everything you need to fix the issue with minimal work. If not I'll be glad to help.

Best regards.

**Comment by zeniko** [comment]
Thanks for the suggestion. Does commit 91c3eb501dcd7a0aacf5c181c6cfadb0c2c787aa work for you?

**Comment by selmf** [comment]
Thanks for the reply!

It works if you add #include <unistd.h> to stream.c. This needs to be done because the _POSIX_C_SOURCE macro is undefined otherwise, which will disable the code.

http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/unistd.h.html

It probably would be best to wrap the include in some reliable unix detection macros, maybe something like this:

#if !defined(_WIN32) && (defined(__unix__) || defined(**unix) || (defined(__APPLE**) && defined(**MACH**)))
#include <unistd.h>
#endif

**Comment by zeniko** [comment]
As I read http://pubs.opengroup.org/onlinepubs/007904975/functions/xsh_chap02_02.html , `_POSIX_C_SOURCE` must be defined before the inclusion of any header. Consequently it doesn't seem to be defined in unistd.h after all.

I've tried this in my Debian installation and it worked as expected. Apparently, gcc only defines `_POSIX_C_SOURCE` when -std=c99 is _not_ set, though. Have you tried compiling with the included Makefile or your CMake script? If the latter, does explicitly defining _POSIX_C_SOURCE (with value 200112L) help?

**Comment by selmf** [comment]
Yes, as far as I know using -std=c99 disables some features for better ansi C compatibility. I removed it from the CMake script when I started experimenting with fseeko and ftello, since those are not ansi functions and gcc will complain about that in c99 mode.

I'm not sure what went wrong on my side. When I tested both your Makefile and the CMake script two days ago the macro for some reason didn't get activated and unistd.h seemed to fix that issue.

Now, two days later, I reran the same test (while making sure unistd.h is absent), and I get the same result as you. It just works, both your Makefile and my CMake script. I guess I either forgot to properly clean my build environment or it's a case of Vorführeffekt...

So, I'm sorry for the extra effort I had you go through.

Explicitily defining the macro is a good idea. I'm not sure if it is needed, but it sure won't hurt. An alternative would be defining _XOPEN_SOURCE = 600 like it is mentioned in the article you linked.

By the way, I've started shipping the shared library version of unarr with the linux version of "YACReader" (www.yacreader.com). For this, I've created packages for Debian, Ubuntu and Archlinux. So far the users seem to like it :)
